librz/io/p/io_srec.c

```diff
@@ -0,0 +1,797 @@
+				}
+				cksum += byte;
+			}
+			sec_size += bc;
+			next_addr += bc;
+			if (eol) {
+				// checksum
+				if (sscanf(str + 8 + (i * 2), "%02x", &byte) != 1) {
+					eprintf("unparsable data !\n");
+					goto fail;
+				}
+				cksum = ~cksum; //this might be useless cause cksum is ut8
+				if (cksum != byte) {
+					eprintf("Checksum failed (got %02x expected %02x)\n", byte, cksum);
+					goto fail;
+				}
+				*eol = 'S';
+			}
+			str = eol;
+			break;
+
+		case 1: // 16 bit adress, with data field!, kinda same with S0
+
+			// S / 1 / (addres+data+checksum) bytes / 0000 / data / checksum
+			l = sscanf(str + 4, "%04x", &addr_tmp);
+			addr_tmp &= 0xffff;
+
+			eol = strchr(str + 1, 'S');
+			if (eol) {
+				*eol = 0;
+			}
+
+			cksum = bc;
+			cksum += addr_tmp >> 8;
+			cksum += addr_tmp & 0xff;
+
+			if ((next_addr != addr_tmp) || ((sec_size + bc) > SEC_MAX)) {
+				// previous block is not contiguous, or
+				// section buffer is full => write a sparse chunk
+				if (sec_size && sec_size < UT16_MAX) {
+					if (rz_buf_write_at(rbuf, sec_start, sec_tmp, (st32)sec_size) != sec_size) {
+						eprintf("sparse buffer problem, giving up\n");
+						goto fail;
+					}
+				}
+				// advance cursor, reset section
+				sec_start = segreg + addr_tmp;
+				next_addr = addr_tmp;
+				sec_size = 0;
+			}
+
+			for (i = 0; i < bc - 3; i++) {
+				if (sscanf(str + 8 + (i * 2), "%02x", &byte) != 1) {
+					eprintf("unparsable data !\n");
+					goto fail;
+				}
+				if (sec_size + i < sec_count) {
+					sec_tmp[sec_size + i] = (ut8)byte & 0xff;
+				}
+				cksum += byte;
+			}
+			sec_size += bc;
+			next_addr += bc;
+			if (eol) {
+				// checksum
+				if (sscanf(str + 8 + (i * 2), "%02x", &byte) != 1) {
+					eprintf("unparsable data !\n");
+					goto fail;
+				}
+				cksum = ~cksum;
+				if (cksum != byte) {
+					eprintf("Checksum failed (got %02x expected %02x)\n", byte, cksum);
+					goto fail;
+				}
+				*eol = 'S';
+			}
+			str = eol;
+			break;
+
+		case 5: // optional, the adress field is a 16bit counter of S1/S2/S3 records
+
+			// S / 5 / (counter+checksum) bytes (3)/ counter / - / checksum
+
+			l = sscanf(str + 4, "%04x", &addr_tmp);
+			addr_tmp &= 0xffff;
+
+			eol = strchr(str + 1, 'S');
+			if (eol) {
+				*eol = 0;
+			}
+
+			cksum = bc;
+			cksum += addr_tmp >> 8;
+			cksum += addr_tmp & 0xff;
+
+			if ((next_addr != addr_tmp) || ((sec_size + bc) > SEC_MAX)) {
+				// previous block is not contiguous, or
+				// section buffer is full => write a sparse chunk
+				if (sec_size && sec_size < UT16_MAX) {
+					if (rz_buf_write_at(rbuf, sec_start, sec_tmp, (st32)sec_size) != sec_size) {
+						eprintf("sparse buffer problem, giving up\n");
+						goto fail;
+					}
+				}
+				// advance cursor, reset section
+				sec_start = segreg + addr_tmp;
+				next_addr = addr_tmp;
+				sec_size = 0;
+			}
+
+			sec_size += bc;
+			next_addr += bc;
+			if (eol) {
+				// checksum
+				if (sscanf(str + 8 + (i * 2), "%02x", &byte) != 1) {
```

Comment:
this `i` may not be initialized.

did you forget to put a for loop here?

Please also add a test for each `S<n>` case

Reply:
No, I don't think there is a need for a `for` loop here because there is no data, so you add only the address bytes for the `checksum`. Same goes for the other cases where there is no data.

Reply:
I think the `i` is the problem indeed. If the first case is one which has no `for` loop, it will not be initialized. I will think about it tonight.

Reply:
Also I will update the test file to cover every case.